pandas/io/formats/csvs.py

```diff
@@ -75,50 +83,92 @@ def __init__(
         self.should_close = ioargs.should_close
         self.mode = ioargs.mode
 
+        # GH21227 internal compression is not used for non-binary handles.
```

Comment:
shouldn't this actually be in `get_filepath_or_buffer` instead? in `pandas/io/common.py`?

cc @twoertwein

Reply:
that is currently not in `get_filepath_or_buffer` (but it would be a good place for it!). We have the same code also in `to_csv`: https://github.com/pandas-dev/pandas/blob/bdb6e26feb26931377a92da1653ff4b95af6351d/pandas/io/formats/csvs.py#L162
If I didn't miss a function, all callers of `get_filepath_or_buffer` use the returned compression from it (if they care about compression).

Reply:
I moved this check on compression method into ``pandas/io/common.py``. See ca888c106.

Reply:
thank you! With this change, I think we don't need the warning in `pandas/pandas/io/formats/csvs.py` line 162 anymore.

`get_filepath_or_buffer` opens some URL-like strings in binary mode (because pandas supports more operations in binary mode) even if the caller specified a non-binary mode (unless the caller explicitly specified text mode with `"t"`). If `(isinstance(filepath_or_buffer, str) and is_url(filepath_or_buffer)) or is_fsspec_url(filepath_or_buffer)`, you will need to use `fsspec_mode` instead of `mode`. I'm not sure how keen the pandas devs are on function decorates, I can imagine that it might be easier to first call `get_filepath_or_buffer` (it will decide whether to use binary/text mode) and then throw the warning.

Reply:
@twoertwein, indeed I removed the warning from ``pandas/io/formats/csvs.py``. All of these checks are not handed to the function ``get_filepath_or_buffer``.
I am not sure I understood: is there anything else to be done on the topic? Because it seems that everything is fine now: tests pass and there is no check on the compression method in ``CSVFormatter``.

Reply:
I'm surprised that it doesn't break this test https://github.com/pandas-dev/pandas/blob/c688a0f4e39f362d1d88c35ab6093bc7963fc525/pandas/tests/io/test_gcs.py#L78 
It should call `get_filepath_or_buffer` with `mode='w'` or `mode='r'` but `get_filepath_or_buffer` will use `fsspec_mode` which adds the binary flag.